File: Fsharp/Products/ProductCatalog.Domain/Product.fs
namespace ProductCatalog.Domain

open System
open Utils

type Dimension = {
    Length: PositiveInt
    Width: PositiveInt
    Height: PositiveInt
}

type Weight = private Weight of PositiveInt
module Weight =
    open PositiveInt
    
    let private Kilo = TryCreate 1000
    
    let FromGrams grams = Weight grams
    let FromKg kg = Weight <| kg * Kilo
    

type ProductId = ProductId of Guid

[<NoEquality; NoComparison>]
type Product =
    { Id: ProductId
      Name: NonEmptyString
      Description: NonEmptyString
      Dimension: Dimension
      Weight: Weight
      Categories: NonEmptyList<CategoryId> }

File: Csharp/ProductCatalog/ProductCatalog.Hexagon/Products/Aggregate/Product.cs
using System.Linq;
using ProductCatalog.Hexagon.Categories;
using ProductCatalog.Hexagon.Categories.Aggregate;
using Shared.Core;
using Shared.Core.DomainModeling;
using Shared.Core.Extensions;
using Shared.Domain;

namespace ProductCatalog.Hexagon.Products.Aggregate
{
    public class Product : AggregateRoot<ProductId>
    {
        private NonEmptyList<CategoryId> _categoryIds = default!;
        
        public Product(
            ProductId id,
            ProductName name,
            ProductDescription description,
            Dimension dimension,
            Weight weight,
            NonEmptyList<CategoryId> categoryIds)
            : base(id)
        {
            Name = name;
            Description = description;
            Dimension = dimension;
            Weight = weight;
            CategoryIds = categoryIds;
        }
        
        public ProductName Name { get; set; }
        public ProductDescription Description { get; set; }
        public Dimension Dimension { get; set; }
        public Weight Weight { get; set; }

        public NonEmptyList<CategoryId> CategoryIds
        {
            get => _categoryIds;
            set =>
                _categoryIds =
                    value
                        .Distinct()
                        .ToNonEmptyList();
        }
    }
}